File: sdk/py/hop_top_kit/output.py
"""
hop_top_kit.output — table/json/yaml renderer.

Mirrors the Go output.Render API from hop.top/kit/go/console/output/renderer.go.

Public surface
--------------
Format  — Literal type alias: 'table' | 'json' | 'yaml'
render  — write *v* to *w* in the requested *format*

Table alignment algorithm
-------------------------
1. Collect all rows as lists of str values.
2. Compute ``col_widths[i] = max(len(header[i]), max(len(row[i]) for row))``
   for each column index *i*.
3. Pad every cell to its column width with trailing spaces.
4. Join cells in each row with a 2-space gap ('  ').
5. Write header row, then each data row, each terminated with '\\n'.
6. Empty input list → no output (not even a header).

Usage
-----
>>> import io
>>> from hop_top_kit.output import render
>>> buf = io.StringIO()
>>> render(buf, 'json', {'key': 'val'})
>>> print(buf.getvalue())
{
  "key": "val"
}
"""

from __future__ import annotations

import dataclasses
import json
from typing import IO, Any, Literal

import yaml

Format = Literal["table", "json", "yaml"]

_GAP = "  "  # 2-space column separator


# ---------------------------------------------------------------------------
# Public
# ---------------------------------------------------------------------------


def render(w: IO[str], format: Format, v: Any) -> None:
    """Write *v* to *w* in *format*.

    Parameters
    ----------
    w:
        Any writable text stream (e.g. ``sys.stdout``, ``io.StringIO``).
    format:
        One of ``'json'``, ``'yaml'``, or ``'table'``.
    v:
        Value to render.  For ``'table'`` this must be a list of dicts,
        a list of dataclasses, a single dict, or a single dataclass.
        For ``'json'`` / ``'yaml'`` any JSON/YAML-serialisable value works.

    Raises
    ------
    ValueError
        If *format* is not one of the three recognised values.
    """
    if format == "json":
        w.write(json.dumps(v, indent=2))
        w.write("\n")
    elif format == "yaml":
        w.write(yaml.dump(v))
    elif format == "table":
        _render_table(w, v)
    else:
        raise ValueError(f"unknown output format {format!r} (valid: json, yaml, table)")


# ---------------------------------------------------------------------------
# Table helpers
# ---------------------------------------------------------------------------


def _to_rows(v: Any) -> tuple[list[str], list[list[str]]]:
    """Return ``(headers, data_rows)`` from *v*.

    Handles:
    - list of dicts
    - list of dataclasses
    - single dict
    - single dataclass

    An empty list returns ``([], [])``.
    """
    if isinstance(v, list):
        if not v:
            return [], []
        first = v[0]
        if dataclasses.is_dataclass(first) and not isinstance(first, type):
            headers = [f.name for f in dataclasses.fields(first)]
            rows = [[str(getattr(item, h)) for h in headers] for item in v]
        else:
            # list of dicts (or dict-like)
            headers = list(first.keys())
            rows = [[str(item[h]) for h in headers] for item in v]
        return headers, rows

    if dataclasses.is_dataclass(v) and not isinstance(v, type):
        headers = [f.name for f in dataclasses.fields(v)]
        return headers, [[str(getattr(v, h)) for h in headers]]

    if isinstance(v, dict):
        headers = list(v.keys())
        return headers, [[str(v[h]) for h in headers]]

    raise TypeError(f"render table: unsupported type {type(v)!r}")


def _render_table(w: IO[str], v: Any) -> None:
    """Format and write a table with aligned, 2-space-gapped columns."""
    headers, rows = _to_rows(v)
    if not headers:
        return  # empty list — no output

    # Compute per-column max widths (header participates too).
    col_widths = [len(h) for h in headers]
    for row in rows:
        for i, cell in enumerate(row):
            if len(cell) > col_widths[i]:
                col_widths[i] = len(cell)

    def _fmt_row(cells: list[str]) -> str:
        padded = [cell.ljust(col_widths[i]) for i, cell in enumerate(cells)]
        return _GAP.join(padded).rstrip()

    w.write(_fmt_row(headers) + "\n")
    for row in rows:
        w.write(_fmt_row(row) + "\n")

File: sdk/ts/src/output.ts
/**
 * @module output
 * @package @hop-top/kit
 *
 * Output rendering — writes structured values to a writable stream in one of
 * three formats: table, JSON, or YAML.
 *
 * ## Quick start
 *
 * ```ts
 * import { render, TABLE_FORMAT, JSON_FORMAT, YAML_FORMAT } from '@hop-top/kit/output';
 *
 * // Write a table to stdout
 * render(process.stdout, TABLE_FORMAT, [{ id: '1', name: 'Alice' }]);
 *
 * // Write indented JSON
 * render(process.stdout, JSON_FORMAT, { ok: true });
 *
 * // Write YAML
 * render(process.stdout, YAML_FORMAT, { version: 1 });
 * ```
 *
 * ## Table column alignment algorithm
 *
 * 1. Derive headers from `Object.keys(rows[0])` (or `Object.keys(obj)` for a
 *    single object).
 * 2. For each column compute `maxWidth = max(header.length, max(cell.length))`.
 * 3. Every cell and header is padded with trailing spaces to `maxWidth`.
 * 4. Columns are separated by two spaces (`"  "`).
 * 5. The resulting lines are all the same width, making them easy to scan.
 *
 * An empty array produces **no output** — not even a header row.  Callers
 * that need a "no results" message should check length before calling `render`.
 */

import * as yaml from 'js-yaml';

/**
 * Supported output formats.
 *
 * - `"table"` — aligned ASCII table; headers derived from object keys.
 * - `"json"`  — `JSON.stringify` with 2-space indentation + trailing newline.
 * - `"yaml"`  — YAML serialisation via js-yaml.
 */
export type Format = 'table' | 'json' | 'yaml';

/** Constant for the JSON output format. */
export const JSON_FORMAT = 'json' as const;

/** Constant for the YAML output format. */
export const YAML_FORMAT = 'yaml' as const;

/** Constant for the table output format. */
export const TABLE_FORMAT = 'table' as const;

/**
 * Renders `v` to `w` in the requested `format`.
 *
 * @param w       - Destination writable stream (e.g. `process.stdout`).
 * @param format  - One of {@link JSON_FORMAT}, {@link YAML_FORMAT}, {@link TABLE_FORMAT}.
 * @param v       - Any JSON/YAML-serialisable value.  For table format, pass
 *                  an array of plain objects or a single plain object.
 *
 * @throws {Error} For unknown format strings.
 */
export function render(w: NodeJS.WritableStream, format: Format, v: unknown): void {
  switch (format) {
    case JSON_FORMAT:
      w.write(JSON.stringify(v, null, 2) + '\n');
      break;
    case YAML_FORMAT:
      w.write(yaml.dump(v));
      break;
    case TABLE_FORMAT:
      renderTable(w, v);
      break;
    default: {
      // exhaustive check — TypeScript narrows to `never` here
      const _exhaustive: never = format;
      throw new Error(
        `unknown output format "${_exhaustive}" (valid: json, yaml, table)`,
      );
    }
  }
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/**
 * Renders `v` as an aligned ASCII table.
 *
 * Accepts:
 *  - An array of plain objects → header row + one row per element.
 *  - A single plain object    → header row + one data row.
 *  - An empty array           → no output.
 */
function renderTable(w: NodeJS.WritableStream, v: unknown): void {
  const rows = normaliseRows(v);
  if (rows.length === 0) return;

  const headers = Object.keys(rows[0]);
  const cells: string[][] = rows.map(row =>
    headers.map(h => String((row as Record<string, unknown>)[h] ?? '')),
  );

  // Compute max width per column (header vs all cell values).
  const widths = headers.map((h, ci) =>
    Math.max(h.length, ...cells.map(row => row[ci].length)),
  );

  const pad = (s: string, w: number) => s + ' '.repeat(w - s.length);
  const formatRow = (cols: string[]) =>
    cols.map((c, i) => pad(c, widths[i])).join('  ');

  w.write(formatRow(headers) + '\n');
  for (const row of cells) {
    w.write(formatRow(row) + '\n');
  }
}

/**
 * Normalises `v` into an array of plain objects.
 *
 * - Array input → returned as-is (may be empty).
 * - Single object input → wrapped in a 1-element array.
 */
function normaliseRows(v: unknown): Record<string, unknown>[] {
  if (Array.isArray(v)) {
    return v as Record<string, unknown>[];
  }
  return [v as Record<string, unknown>];
}
